### PR TITLE
Improving Progressive Web App offline support detection

### PR DESCRIPTION
**Pull request by petele**
Changes proposed in this pull request:
- PWA offline detection changes post
- Adds tag for `progressive-web-apps`

**TODO Before Merging:**

* [ ] Update shortlink https://goo.gle/improved-pwa-offline-detection 

cc: @b1tr0t

### CODE REVIEW

**Comment by petele** [comment]
Have LGTM from @b1tr0t via email

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+Only apps that fulfill all core
+[Progressive Web App installability criteria][pwa-criteria], including support
+for an offline mode, can be installed to the device from Chrome.
+
+## Current offline detection logic
+
+Verification of offline support has been part of the browser's installability
+checklist of a PWA for several years. However, the check was a simplified
+heuristic.
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:50]:
```suggestion
Verification of offline support has been part of PWA installability
criteria for several years; however, the check was a simplified
heuristic.
```

Comment (jpmedley) [reply site/en/blog/improved-pwa-offline-detection/index.md:50]: Also, should it be "Chrome's PWA installability criteria"?

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+Previously, the installability check passed if the page had a service worker
+installed that included a `fetch` event handler.  Chrome did not have the
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:53]:
```suggestion
that included a `fetch` event handler.  Chrome did not have the
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+capability to simulate requests through the service worker, so a full check of
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:54]:
```suggestion
ability to simulate requests through the service worker, so a full check of
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+correct offline behaviour was not possible.
+
+{% img src="image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/qrDoFGCM2s8pBvM4n7Dv.png", alt="Diagram of service worker", width="800", height="385" %}
+
+That meant that Chrome did not have the ability to validate whether the `fetch`
+event handler returned a valid resource with HTTP 200 during the offline check.
+Chrome only checked whether the service worker actually has a `fetch` handler.
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:61]:
```suggestion
Chrome only checked whether the service worker actually had a `fetch` handler.
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+Starting in Chrome 89, Chrome has the ability to run simulated offline requests
+through the service worker, allowing us to update the offline detection logic
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:64]:
```suggestion
through the service worker, allowing improved offline detection logic
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+to better reflect actual offline support of the application.
+
+## Updated offline detection logic
+
+The updated offline detection logic checks:
+
+1) That there is a service worker installed for the page.
+2) That the installed service worker has a `fetch` event.
+3) **That the installed service worker `fetch` event returns an HTTP 200
+   status code (indicating a successful fetch) in simulated offline mode.**
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:74]:
```suggestion
* That there is a service worker installed for the page.
* That the installed service worker has a `fetch` event.
* **That the installed service worker `fetch` event returns an HTTP 200
   status code (indicating a successful fetch) in simulated offline mode.**
```

Comment (jpmedley) [reply site/en/blog/improved-pwa-offline-detection/index.md:74]: Web.dev restricts numbered lists to steps of a procedure.

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+## What does this mean for developers?
+
+A properly implemented offline mode requires that the `fetch` handler return
+a resource with HTTP 200, so, if you've correctly implemented an offline mode
+for your Progressive Web App, this change to the offline detection logic will
+not require any action on your part.
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:81]:
```suggestion
for your Progressive Web App, this change to offline detection requires no
action by you.
```

Comment (jpmedley) [reply site/en/blog/improved-pwa-offline-detection/index.md:81]: Aside from following a best practice of omitting needless words, this makes a primary point of the article more direct, and hence easier to remember.

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+offline mode isn't working properly and a resource is not returned with an
+HTTP 200 response code, you will need to update your PWA to meet the updated
+installability criteria.
+
+It's up to you to decide what kind of offline experience you want to provide.
+On one end of the spectrum is a fully functional offline experience. This means
+pre-caching all the resources and data needed, and syncing data with your
+server when the user is online again. Caching resources will also help improve
+[core web vital metrics][cwv] because it eliminates the need to download
+resources from the network every time. At the other end of the spectrum, a
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:93]:
```suggestion
resources from the network every time. At the other end of the spectrum is a
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+[custom offline fallback page][offline-fallback].
+
+!!!.aside
+Check out [Workbox][workbox], a set of libraries that can power a
+production-ready service worker for your Progressive Web App.
+
+**Have a question about service workers?** Ask it on [Stack Overflow][so] and
+tag it with [`service-worker`][so-sw] and [`progressive-web-apps`][so-pwa],
+our team regularly monitors those tags and does our best to help.
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:102]:
```suggestion
tag it with [`service-worker`][so-sw] and [`progressive-web-apps`][so-pwa].
Our team regularly monitors those tags and does our best to help.
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+!!!
+
+## When will this change take effect?
+
+For most PWA developers who correctly implemented offline mode, this change
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:107]:
```suggestion
For most PWA developers, specifically those who correctly implemented offline mode, this change
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+will not impact your applications.  However, we want to give adequate time
+for developers with apps that have broken offline modes to address the issue.
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:109]:
```suggestion
to address this issue for developers with apps that have broken offline modes.
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+### Warning starting Chrome 89 (March 2021)
+
+Starting in Chrome 89 (stable in March 2021), if the PWA does not provide a
+valid response when offline, a message will be displayed under the *Issues*
+tab of developer tools. The `beforeinstallprompt` event, and in-browser
+installation prompts will still be provided.
+
+Lighthouse v7 will also indicate that there's a problem with the
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:118]:
```suggestion
Lighthouse v7 and later will also indicate when there's a problem with the
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+offline implementation and provide guidance on how to fix it.
+
+<figure>
+  {% img src="image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/PuyRpRfJnVs8o8aTdHlA.png", alt="Screen shot of DevTools showing warning message in Console", width="800", height="115" %}
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:122]:
```suggestion
  {% img src="image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/PuyRpRfJnVs8o8aTdHlA.png", alt="Screen shot of DevTools showing warning message in Console.", width="800", height="115" %}
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  <figcaption>
+    Warning message in the Chrome DevTools Console
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:124]:
```suggestion
    Warning message in the Chrome DevTools Console.
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+  </figcaption>
+</figure>
+
+<figure>
+  {% img src="image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/TjqKPXYUF5Y3tjeZLSwT.png", alt="Screenshot of DevTools showing warning message in Application tab", width="800", height="175" %}
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:129]:
```suggestion
  {% img src="image/0g2WvpbGRGdVs0aAPc6ObG7gkud2/TjqKPXYUF5Y3tjeZLSwT.png", alt="Screenshot of DevTools showing warning message in Application tab.", width="800", height="175" %}
```

File: site/en/blog/improved-pwa-offline-detection/index.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+  <figcaption>
+    Warning message in Application tab &gt; Manifest &gt; Installability
```
Comment (jpmedley) [inline site/en/blog/improved-pwa-offline-detection/index.md:131]:
```suggestion
    Warning message in Application tab &gt; Manifest &gt; Installability.
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

**Comment by petele** [comment]
@jpmedley - I've integrated all of your feedback.

**Review by jpmedley — APPROVED** [review]

**Comment by petele** [comment]
Please - **do not merge** - I'd like to personally merge this and push it out with some specific tweets and other announcements.

**Comment by jpmedley** [comment]
The new rule for like the last month has been for authors to merge their own work.

**Comment by jpmedley** [comment]
@robdodson Percy is indicating something needs to be reviewed. When I look at the details, I don't know what I'm supposed to be looking at or even if I'm the one supposed to be doing it. I'd appreciate a little guidance.

**Comment by robdodson** [comment]
Should be fixed now. Currently percy is snapshotting live data so basically every new post will cause it to fail (because the page always changes). https://github.com/GoogleChrome/developer.chrome.com/issues/78

We'll fix the bug and write up a better explainer soon. For now feel free to tag me when percy fails and I can approve.

**Review by robdodson — APPROVED** [review]

**Review by jpmedley — APPROVED** [review]

**Pull request merged**